### PR TITLE
Add a NewConfigProvider that uses options for settings, and does input validation

### PR DESCRIPTION
**Pull request by bogdandrutu**
This is a fork of https://github.com/open-telemetry/opentelemetry-collector/pull/4762

Also, deprecates `service.MustNewConfigProvider` and `service.MustNewDefaultConfigProvider`in favor of `service.NewConfigProvider`.

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

### CODE REVIEW

File: service/collector.go
Comment (bogdandrutu) [reply service/collector.go:53]: Yeah, I made this PR, then I realized that I wanted to cleanup some other things :)) so I created the cleanup PRs.

File: service/config_provider_test.go
Comment (bogdandrutu) [reply service/config_provider_test.go:168]: We usually don't do that...

File: service/config_provider.go
Comment (bogdandrutu) [reply service/config_provider.go]: done.

File: service/config_provider_test.go
Comment (mx-psi) [reply service/config_provider_test.go:168]: That's fair, let's not do it then

File: CHANGELOG.md
Comment (bogdandrutu) [reply CHANGELOG.md]: Thanks for catching this.

File: service/config_provider_test.go
Comment (jpkrohling) [reply service/config_provider_test.go:168]: To be honest, I've been requesting those on the PRs I'm reviewing. I think it's important to assert that the behavior didn't change.

Comment (bogdandrutu) [reply service/config_provider_test.go]: Done.

File: service/config_provider.go
Comment (bogdandrutu) [reply service/config_provider.go]: Done...

**Comment by bogdandrutu** [comment]
@open-telemetry/collector-approvers PTAL

File: service/config_provider.go
Comment (Aneurysm9) [inline service/config_provider.go:109]:
```suggestion
// Deprecated: [v0.49.0] use NewConfigProvider instead.
```

File: service/config_provider.go
Comment (Aneurysm9) [inline service/config_provider.go:102]: I'm not sure I understand this, or the comment on the same topic in the changelog.  Can you provide an example here of how it would be done?  It seems to me that the `config.Map` has already been turned into a `config.Config` and so there is no ability to influence that translation, only the `config.Map` or the `config.Config`.

This is a capability that exists now.  Are we sure that it is unused and will not present issues if we remove it?  We have made assumptions before about how people use the config processing system and got it very wrong.  I'd like to avoid that happening again.

File: service/config_provider.go
Comment (Aneurysm9) [inline service/config_provider.go:96]:
```suggestion
		opts.configMapConverters = append(opts.configMapConverters, c)
```

This should append rather than replace so that multiple `WithConfigMapConverters` options can be composed.

File: service/config_provider.go
Comment (Aneurysm9) [inline service/config_provider.go:165]:
```suggestion
// Deprecated: [v0.49.0] use NewConfigProvider instead.
```

File: service/config_provider.go
Comment (Aneurysm9) [inline service/config_provider.go:151]: This drops the `NewOverwritePropertiesConverter`.  While equivalent functionality may be available from the `yamlmapprovider` package I would expect this to surprise anyone using the `--set` flag with a suddenly non-functioning deployment.

File: service/config_provider.go
Comment (bogdandrutu) [reply service/config_provider.go:96]: The order of the converters matter, so we cannot do append here because will confuse lots of users. The reason why matters is because I may implement a converter that adds a property using an "envvar" then if the env converter is called before or after produces different results.

File: service/config_provider.go
Comment (bogdandrutu) [reply service/config_provider.go:165]: Done.

File: service/config_provider.go
Comment (bogdandrutu) [reply service/config_provider.go:109]: Done.

File: service/config_provider.go
Comment (bogdandrutu) [reply service/config_provider.go:151]: The `NewOverwritePropertiesConverter` cannot be added here, because it needs the values for `--set`. This is not a regression and not a "non-functioning" deployment, because:
* If users use our distros we take care of that in NewWindowsService and NewCommand;
* If users use a custom distro that does not use NewWindowsService and/or NewCommand and they use:
  *  `MustNewConfigProvider` - that did not have support for `--set` callers needed to specifically use `NewOverwritePropertiesConverter`; When they transition nothing to do;
  *  `MustNewDefaultConfigProvider` that did have support for `--set`; When transition callers need to specifically use `NewOverwritePropertiesConverter` for that;

File: service/config_provider.go
Comment (bogdandrutu) [reply service/config_provider.go:102]: I am 99.99% sure (that the only use-case we heard so far and seen is to alter an unmarshalled `config.Config`, and not to change the unmarshalling logic), that being said, now we have a clear process in place for deprecation. If we make a mistake here, I expect users to come back and tell us that they need that functionality and stop the removal.

I am happy to actually expose this as deprecated, with the "proposed transition", and a note that this will not be removed for at least 2 versions to ensure that possible users are coming and telling us.

Comment (bogdandrutu) [reply service/config_provider.go]: Reverted this, since is a different change. will submit a different PR to argue about this.

File: service/config_provider.go
Comment (mx-psi) [reply service/config_provider.go:96]: >  we cannot do append here because will confuse lots of users. 

Should `NewConfigProvider` return an error if multiple `WithConfigMapConverters` functions are passed then?

Comment (bogdandrutu) [reply service/config_provider.go:96]: Not sure... Personally I don't see a reason to do that.

Comment (bogdandrutu) [reply service/config_provider.go:96]: But also this is complicated. We want to have a `default` for this like the expand converter (which probably should not be a converter but that is a separate discussion), if user overwrites that should be accepted correct? Only if it double overwrites is error? Also API accepts slice, so that forces users to think about this...

If you really think that this should be an error let me know, but I don't see this as an error, also the users of this API are very few people :))

File: CHANGELOG.md
Comment (codeboten) [inline CHANGELOG.md:47]: Seems like a typo here
```suggestion
```

File: service/config_provider_test.go
Comment (codeboten) [inline service/config_provider_test.go:37]:
```suggestion
	scheme string
```

File: service/config_provider_test.go
Comment (codeboten) [inline service/config_provider_test.go:59]:
```suggestion
	if m.scheme == "" {
```

File: service/config_provider_test.go
Comment (codeboten) [inline service/config_provider_test.go:62]:
```suggestion
	return m.scheme
```

File: service/config_provider_test.go
Comment (codeboten) [inline service/config_provider_test.go:103]:
```suggestion
				&mockProvider{scheme: "err", errR: errors.New("retrieve_err")},
```

File: service/config_provider_test.go
Comment (codeboten) [inline service/config_provider_test.go:149]:
```suggestion
					&mockProvider{scheme: "err", retM: cfgMap, errC: errors.New("close_err")},
```

File: service/config_provider_test.go
Comment (bogdandrutu) [reply service/config_provider_test.go:59]: I like how IDE was smart enough so if I wrote a typo once, helped me to survive with that :)))

File: service/config_provider_test.go
Comment (bogdandrutu) [reply service/config_provider_test.go:103]: Done.

File: service/config_provider_test.go
Comment (bogdandrutu) [reply service/config_provider_test.go:149]: Done.

File: service/config_provider_test.go
Comment (bogdandrutu) [reply service/config_provider_test.go:62]: Done.

File: service/config_provider_test.go
Comment (bogdandrutu) [reply service/config_provider_test.go:37]: Done.

File: CHANGELOG.md
Comment (bogdandrutu) [reply CHANGELOG.md]: Done.

**Comment by bogdandrutu** [comment]
@codeboten thanks for the review, PTAL

File: service/config_provider.go
Comment (codeboten) [reply service/config_provider.go:96]: I think the argument being an array makes it clear enough that folks should set multiple with a single call here.

I can see how it's confusing for `WithConfigMapConverters` to overwrite an option each time it's called while `WithConfigMapProvider` appends to the configuration. I don't have a good idea of how to discern between the two other than with documentation.

Maybe `WithConfigMapProvider` should also overwrite the option to remain consistent?

**Review by codeboten — COMMENTED** [review]
Thanks for addressing my suggestions. Just one comment in the discussion w/ @Aneurysm9 & @mx-psi. It would be good to agree on the way to move forward.

File: service/config_provider.go
Comment (Aneurysm9) [reply service/config_provider.go:96]: `WithConfigMapProvider` modifies a map, though, where order does not matter and multiple providers may come from multiple sources.  If anything, I'd like to see `WithConfigMapConverters` move in that direction, creating a set of available converters that could then be referenced in a list of conversion steps to apply.  I don't know that there's anything that can be done one way but not the other, perhaps with the exception of having a component with a single method that returns a slice of options intended to register providers and converters.

Comment (bogdandrutu) [reply service/config_provider.go:96]: The provider is only one so not sure... We can accept a slice there as well. Also the providers are called in the order of the location argument so their order of registration does not matter.

Another option is to have only one Converter, and add support for MultiConverter as a helper implementation, and then that will overwrite the default converter again

File: service/config_provider.go
Comment (Aneurysm9) [reply service/config_provider.go:149]: We need to commit to maintaining support for `--set`, then, because removal of that flag will break existing deployments.  We cannot decide that having two ways to do the same thing is wrong and remove it.

Comment (bogdandrutu) [reply service/config_provider.go:149]: Who we?

Comment (bogdandrutu) [reply service/config_provider.go:149]: The flag is not removed as pointed. Not sure what you are talking, sorry.

Comment (bogdandrutu) [reply service/config_provider.go:149]: How is this discussion relevant to this PR?

File: service/config_provider.go
Comment (bogdandrutu) [reply service/config_provider.go:96]: @codeboten maybe another good option is o have a "ConfigProviderSettings" instead of using options. And would be clear that we have a map there for providers and a slice for converters. Users can see them easily then?

Comment (codeboten) [reply service/config_provider.go:96]: I think that would work. 👍

Comment (bogdandrutu) [reply service/config_provider.go:96]: Will give a try in a different PR or commit

Comment (jpkrohling) [reply service/config_provider.go:96]: I think the Settings solution is the best solution to solve the ambiguity that multiple `WithConfigMapConverters` options would bring.

File: service/config_provider.go
Comment (jpkrohling) [inline service/config_provider.go:172]: I guess this is so early in the code that we don't have even a basic logging facility set, right?

**Review by jpkrohling — COMMENTED** [review]
LGTM, the only contention point seems to be about calling WithConfigMapConverters multiple times. Once that is sorted out, either by returning an error on the second call or by using a settings object instead of options, this would be ready to be merged.

File: service/config_provider.go
Comment (bogdandrutu) [reply service/config_provider.go:172]: Yes

File: service/collector_test.go
Comment (jpkrohling) [inline service/collector_test.go:111]: I didn't know this before, but I prefer this so much more than the previous iteration.

File: service/config_provider.go
Comment (bogdandrutu) [reply service/config_provider.go]: Done, settings is available now.

**Review by jpkrohling — APPROVED** [review]

File: service/config_provider.go
Comment (codeboten) [inline service/config_provider.go:94]:
```suggestion
	// Unmarshaler is the configunmarshaler.ConfigUnmarshaler to be used to unmarshal the config.Map into config.Config.
```

File: service/config_provider.go
Comment (codeboten) [inline service/config_provider.go:95]:
```suggestion
	// It is required to not be nil, use configunmarshaler.NewDefault() by default.
```

File: service/config_provider.go
Comment (codeboten) [inline service/config_provider.go:99]: I'm probably missing this somewhere, but as a user is there a way for me to obtain the defaults from outside this package? I'm thinking of the case where someone may only want to override the locations but not the default providers/converters

**Review by codeboten — COMMENTED** [review]
Overall I like this a lot. Just a question and a couple of nits

File: service/config_provider.go
Comment (bogdandrutu) [reply service/config_provider.go:99]: I think, if we go with the same model as the Factories for components, we should not have any "defaults" every distribution will pick their "defaults" that they commit to support. All the components that we use in the "default" config are also public, so a distro can emulate the same if this is what they want.

**Review by codeboten — APPROVED** [review]

File: service/config_provider.go
Comment (mx-psi) [reply service/config_provider.go:96]: This still needs to be applied

File: service/command.go
Comment (mx-psi) [inline service/command.go:44]: This and the `collector_windows` code are the same, should we extract it to a function?

**Review by mx-psi — APPROVED** [review]
Using a struct resolves the ambiguity of repeating an option, I only have two nits but otherwise LGTM

**Pull request merged**
